### PR TITLE
[Metrics] Refactored metrics exporting to properly handle high-cardinality cases

### LINKED ISSUE #40033: [Core] Metrics being lost for actors/tasks
### What happened + What you expected to happen

We have a Serve application deployed and periodically we see metrics vanish for some deployments after a while (see screenshot)

Screenshot 2023-10-02 at 5 52 19 PM

### Versions / Dependencies

Ray 2.7

### Reproduction script

N/A

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by alexeykudinkin**
## Why are these changes needed?

Problem
----

We have a use-case where some of our metrics have very high-cardinality (> 60k time-series per metrics), and current implementation unfortunately fails to export metrics reliably in that case due to:

 - Current implementation counts `Metric` proto objects rather than `TimeSeries` towards the batch size limit
 - Eventually once it comes to ~50k time-series we cross default threshold of 20Mb max payload size for gRPC, failing to export metrics and eventually losing them

Solution
----

I'm addressing this by refactoring exporting sequence:

 - Now instead of counting metrics, it'll be appropriately accounting for time-series added into the batch
 - Requests to an agent are now flushed while filling in the time-series, allowing to break up metrics w/ the large number of time-series across multiple requests to an agent

Additionally, to protect from metric loss i'm adding iterative proto payload size checking which is:
 - Checking payload size log(N) times (at 1/2, 3/4, 7/8, ... of the batch size)
 - Validating that the batch stays w/in the max gRPC message size limits set (currently 20Mb)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rynewang** [comment]
I guess we can also split the protos into smaller sized chunks and send the chunks one by one?

**Comment by alexeykudinkin** [comment]
@rynewang that's exactly what this PR is doing

**Review by rkooo567 — COMMENTED** [review]
Change LGTM. There seems to be some failures.

Also since ByteSizeLong is pretty expensive, I wonder if we can do at even at higher granulaity than 25 (100 or 200)? Do you happen to know when you observed the failure, how many metrics were there?

File: src/ray/stats/metric_exporter.cc [hunk 1]
```diff
@@ -161,13 +161,19 @@ void OpenCensusProtoExporter::ExportViewData(
...
+  // TODO define AGENT_GRPC_MAX_MESSAGE_LENGTH to keep it consistent b/w cpp/python
+  size_t binary_size_payload_threshold = (size_t) (AGENT_GRPC_MAX_MESSAGE_LENGTH * .95f);
 
   for (const auto &datum : data) {
     UpdateMetricsData(datum, request_proto);
     data_batched += 1;
-
+    // NOTE: Because each payload size check is linear in the number of fields w/in the payload
+    //       we intentionally sample it to happen only every 25 iterations to avoid affecting performance
+    bool should_check_payload_size = data_batched % 25 == 0
```
Comment (rkooo567) [inline src/ray/stats/metric_exporter.cc:174]: How did you decide 25 here btw? Is it just an arbitrary number?

Comment (alexeykudinkin) [reply src/ray/stats/metric_exporter.cc]: Bumped it up to 1000. Picked it based on my (pseudo-) scientific guesstimate

**Review by rkooo567 — APPROVED** [review]
Awesome PR! Btw, do you think we can test per-series limit inside `test_metrics_batch`? I think we can export a couple of multi-tag metrics and verify it is importing data correctly?

File: python/ray/_private/ray_constants.py [hunk 2]
```diff
@@ -371,7 +371,11 @@ def env_set_by_user(key):
 # The default gRPC max message size is 4 MiB, we use a larger number of 250 MiB
 # NOTE: This is equal to the C++ limit of (RAY_CONFIG::max_grpc_message_size)
 GRPC_CPP_MAX_MESSAGE_SIZE = 250 * 1024 * 1024
+
 # The gRPC send & receive max length for "dashboard agent" server.
+# NOTE: This is equal to the C++ limit of RayConfig::max_grpc_message_size
```
[the remark this answers is not in the document]
Comment (alexeykudinkin) [reply python/ray/_private/ray_constants.py:376]: Should i remove the comment? Or update the value?

**Comment by alexeykudinkin** [comment]
>Awesome PR! Btw, do you think we can test per-series limit inside test_metrics_batch? I think we can export a couple of multi-tag metrics and verify it is importing data correctly?

Doesn't seem like this test exists

**Comment by rkooo567** [comment]
Screenshot 2023-11-20 at 10 06 42 PM

Hmm are you on the right branch?

**Comment by rkooo567** [comment]
We should merge this by Ray 2.9

**Comment by rkooo567** [comment]
lint faliure btw

**Comment by rkooo567** [comment]
@alexeykudinkin the PR description is equivalent to the previous fix I remember. Did you just do refactoring + add tests or added additional functionalities here?

File: src/ray/stats/metric_exporter.h [hunk 3]
```diff
@@ -100,38 +102,83 @@ class OpenCensusProtoExporter final : public opencensus::stats::StatsExporter::H
...
+  /// \param next_payload_size_check_at next batch size when payload (binary) size have
+  /// to be checked
+  void ProcessMetricsData(const opencensus::stats::ViewDescriptor &view_descriptor,
+                          const opencensus::stats::ViewData &view_data,
+                          rpc::ReportOCMetricsRequest &request_proto,
+                          size_t &cur_batch_size,
+                          size_t &next_payload_size_check_at);
+
+ protected:
+  rpc::ReportOCMetricsRequest createRequestProtoPayload();
```
Comment (rkooo567) [inline src/ray/stats/metric_exporter.h:160]: can you add docstring for these methods?

File: src/ray/stats/metric_exporter.h [hunk 4]
```diff
@@ -100,38 +102,83 @@ class OpenCensusProtoExporter final : public opencensus::stats::StatsExporter::H
...
-                                opencensus::stats::ViewData> &datum, rpc::ReportOCMetricsRequest &request_proto);
 
  private:
-  /// Call Manager for gRPC client.
-  rpc::ClientCallManager client_call_manager_;
   /// Lock to protect the client
   mutable absl::Mutex mu_;
   /// Client to call a metrics agent gRPC server.
-  std::unique_ptr<rpc::MetricsAgentClient> client_ ABSL_GUARDED_BY(&mu_);
+  std::shared_ptr<rpc::MetricsAgentClient> client_ ABSL_GUARDED_BY(&mu_);
```
Comment (rkooo567) [inline src/ray/stats/metric_exporter.h:172]: any reason why this should be a shared pointer? IIUC this is not used except within metrics_Exporter? (is it for testing?)

File: src/ray/stats/metric_exporter.h [hunk 5]
```diff
@@ -100,38 +102,83 @@ class OpenCensusProtoExporter final : public opencensus::stats::StatsExporter::H
...
   /// The worker ID of the current component.
   WorkerID worker_id_;
   /// The maximum batch size to be included in a single gRPC metrics report request.
   size_t report_batch_size_;
+  /// Proto request payload size threshold upon reaching which request have to be flushed
+  /// immediately, so that we can make sure that batches stay w/in the threshold of the
+  /// gRPC max message size set by an agent (usually calculated as 95% of agent's gRPC
+  /// max-message size)
+  size_t proto_payload_size_threshold_;
```
Comment (rkooo567) [inline src/ray/stats/metric_exporter.h:181]: add units? (proto_payload_size_threshold_bytes_)

File: src/ray/stats/metric_exporter_grpc_test.cc [hunk 6]
```diff
@@ -149,6 +144,278 @@ TEST(MetricPointExporterTest, adds_global_tags_to_grpc) {
...
+      // Each of the batches have to have 1 metric with 2 time-series each
+      auto metrics = requests[i].metrics();
+      // ASSERT_THAT(metrics.size(), 1);
+      ASSERT_THAT(metrics[0].timeseries().size(), 2);
+    }
+  }
+
+  {
+    //
+    // Test #3: Splitting time-series across 1 batches (no overflowing)
```
Comment (rkooo567) [inline src/ray/stats/metric_exporter_grpc_test.cc:381]: do we have a test case where there's overflow of size and also batches? If not, can you add a simple one?

**Review by rkooo567 — APPROVED** [review]
nit comments!

File: src/ray/stats/metric_exporter.h (hunk 3 above)
Comment (alexeykudinkin) [reply src/ray/stats/metric_exporter.h:160]: These are internal methods (abstracting common boiler-plate) that are not meant to be used externally and on top of it these are also quite straightforward so not sure if it's worthwhile to add doc-comments for them

File: src/ray/stats/metric_exporter.h (hunk 4 above)
Comment (alexeykudinkin) [reply src/ray/stats/metric_exporter.h:172]: Yes, this is needed for testing

File: src/ray/stats/metric_exporter.h (hunk 5 above)
Comment (alexeykudinkin) [reply src/ray/stats/metric_exporter.h:181]: Good call!

File: src/ray/stats/metric_exporter_grpc_test.cc (hunk 6 above)
Comment (alexeykudinkin) [reply src/ray/stats/metric_exporter_grpc_test.cc:381]: Yes, so both cases are covered in `export_view_data_split_by_batch_size` and `export_view_data_split_by_payload_size` respectively

**Comment by rkooo567** [comment]
lmk when it is ready to merge

**Comment by alexeykudinkin** [comment]
@rkooo567 it's ready to merge, getting some weird build failures unrelated to the change

**Comment by alexeykudinkin** [comment]
Mac builds are failing b/c we're trying to upload too much to BK:

Screenshot 2023-11-22 at 10 32 17 AM

**Pull request merged**